### PR TITLE
Loader to use name as identifier rather than url

### PR DESCRIPTION
**Pull request by hypersurge**
using resource name (rather than url) as unique identifier allows alternative loaders to preload the assets (e.g. with decaching query strings)

### CODE REVIEW

File: src/mesh/mesh.ts [hunk 1]
```diff
@@ -173,4 +174,13 @@ export class Mesh3D extends Container3D {
   static createPlane(material: Material = new StandardMaterial()) {
     return new Mesh3D(PlaneGeometry.create(), material)
   }
+
+  /**
+   * Creates a new sphere mesh with the specified material.
+   * @param material The material to use.
+   */
+  static createSphere(material?: Material, radius = 1, widthSegments = 32, heightSegments = 16, phiStart = 0, phiLength = Math.PI * 2, thetaStart = 0, thetaLength = Math.PI) {
```
Comment (jnsmalm) [inline src/mesh/mesh.ts:182]: I think it would be better if everything after material would be inside an options object instead. Also add docs for those options.

**Comment by jnsmalm** [comment]
> using resource name (rather than url) as unique identifier allows alternative loaders to preload the assets (e.g. with decaching query strings)

Could you please provide an example so I can understand a bit more?

File: src/mesh/mesh.ts (hunk 1 above)
Comment (jnsmalm) [reply src/mesh/mesh.ts:182]: An interface could be created for that options object.

**Comment by hypersurge** [comment]
Example:
Let's create or load a Texture Resource prior to Cubemap Resource (if loading, perhaps from a url other than specified in the Cubemap data).  By assigning "name" to Texture Resource it can be easily shared to Cubemap's requirements, whereas "url" is one-to-one and may not be compatible.

Why is this impotant?  Without it:
1) We can't rely on our macro loaders (they scrape contents from an assets folder, and we heavily reuse and combine resources)
2) We can't force decache assets at development runtime (e.g. by appending params / timestamps to urls)

In summary "name" is the unique indentifier, whereas "url" is the location of the resource.  So logical filters should be done on the unique identifier ("name"), which may or may not be the same as location ("url") dependent upon use case.

Hope that helps, I accept our specific use case is probably of minor concern to anyone else so have tried to justify in more abstract terms.

File: src/mesh/mesh.ts (hunk 1 above)
Comment (hypersurge) [reply src/mesh/mesh.ts:189]: added jsdocs
disagree on dedicated options object for this case (interfaces less intuitive for multiple params with non-zero defaults, extra init needed etc).

Comment (jnsmalm) [reply src/mesh/mesh.ts:189]: What do you mean extra init needed? The options object should have the same defaults as the args would.

File: src/loader/gltf-loader.ts [hunk 2]
```diff
@@ -34,14 +34,14 @@ class glTFExternalResourceLoader implements glTFResourceLoader {
...
-      this._loader.add({ parentResource: this._resource, url, onComplete })
+      this._loader.add({ parentResource: this._resource, name: url, url: url, onComplete: onComplete })
     } else if (this._loader.resources[url].data) {
       // The resource already exists, just use that one.
       onComplete(this._loader.resources[url])
     } else {
       // The resource is in queue to be loaded, wait for it.
       let binding = this._loader.onProgress.add((_, resource) => {
-        if (resource.url === url) {
+        if (resource.name === url) {
```
Comment (jnsmalm) [inline src/loader/gltf-loader.ts:44]: Double-checking that this condition is really correct?

File: src/mesh/mesh.ts (hunk 1 above)
Comment (hypersurge) [reply src/mesh/mesh.ts:189]: Show example of non-zero default value Options object within this library.

File: src/loader/gltf-loader.ts (hunk 2 above)
Comment (hypersurge) [reply src/loader/gltf-loader.ts:44]: Good point, it does depend how it got in the queue and whether the expected name convention was assigned at that time (e.g. line 37).  An OR may be more appropriate, as either are likely acceptable outcomes.

if ( resource.url == url || resource.name == url )

File: src/mesh/mesh.ts (hunk 1 above)
Comment (jnsmalm) [reply src/mesh/mesh.ts:189]: Something like this using object spread with default values:

```
interface DoSomethingOptions {
  a?: number,
  b?: string
}

function doSomething(options: DoSomethingOptions = {}) {
  let {a = 1, b = "something"} = options
}
```

Comment (hypersurge) [reply src/mesh/mesh.ts:189]: That's what I mean by an extra init needed.

I see this approach is used in one other place in the library with non-zero values (PostProcessingSprite constructor).  Whereas I also see constructors that adopt params with defaults (e.g. Color) so there is inconsistent implementation.

In this case (a factory method), I don't see value in replacing something solved at language level (default params in a method) with a proprietary approach (options object with init).  Factory methods are already expressive helpers.

Drawbacks to consider: more work to use, more definitions to maintain, more work to port functionality from Three etc, proprietary api style.

Can we agree this is an unnecessary burden?

Comment (jnsmalm) [reply src/mesh/mesh.ts:189]: It's harder to understand this code:

`createSphere(material, 1, 32, 16, 0, Math.PI * 2, 0, Math.PI * 2)`

or this:

`createSphere(material, undefined, undefined, undefined, undefined, undefined, undefined, Math.PI * 2)`

to this:

`createSphere(material, {thetaLength: Math.PI * 2})`

This is the problem when having a large number of arguments in a function. This is hardly a proprietary solution, even PixiJS itself uses this in some places. https://pixijs.download/dev/docs/PIXI.DisplayObject.html#destroy for example.

I don't think an options object should be used for required args though.

Comment (hypersurge) [reply src/mesh/mesh.ts:189]: Harder to understand?  Maybe, maybe not.  Depends how one codes or what IDE one uses.  For example I don't use JS or TS, so it's all moot to me.  I recall many of the Pixi methods have overloading to allow params or options as wanted.

Feel free to enhance / implement it this way if you like, but it's not something I care to spend time on.

I'd rather spend said time porting larger chunks of Three (if adopting Pixi3d).  So we wouldn't want the burden of redesigning Three's interface (which has high adoption) for what is largely a decorative rationalisation on this interface (which has low adoption).  Not intended as a criticism of adoption, just utility of cross usage of tutorials, examples etc.

**Comment by jnsmalm** [comment]
You are free to spend your time however you want 👍

**Comment by hypersurge** [comment]
I will spend another minute of my time highlighting the outcome because I am grateful for the groundwork you have done.  By rejecting this PR you chose to have a 3D lib incapable of creating a Sphere primitive or off-loading Cubemap or Gltf assets because of a religious topic.

By arbitrarily and inconsistently making a non-documented, non-essential approach mandatory for reasons of personal style you close further contribution - forcing a fork is self destructive to an open source project.

A polite reminder - when someone takes an open source project seriously enough to contribute and explain their longer term motivations, the response should always be: cool, let's keep the momentum going :)

**Comment by hypersurge** [comment]
That doesn't relate to the Loader fixes as they have no new API.  Style is subjective, if it was critical then it would be a) documented, b) consistently implemented throughout the lib, c) important enough for everyone to role up their sleeves.

Evaluating, bug fixing, porting, enhancements, highlighting inconsistencies, dialogue - these are all contributions, where's that appreciation you mention?

**Comment by jnsmalm** [comment]
Yes, we can merge the resource stuff. Please create a new PR and also add a test case for the functionality you have changed. You can copy the test from https://github.com/jnsmalm/pixi3d/blob/develop/test/gltf.test.js and change it so it works with your case. Also edit https://github.com/jnsmalm/pixi3d/blob/develop/test/index.js to include the new test. Run tests with npm run test.
